idx/metric_definition.go

```diff
@@ -0,0 +1,546 @@
+package idx
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/binary"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+	"unsafe"
+
+	"github.com/raintank/schema"
+	log "github.com/sirupsen/logrus"
+	"github.com/tinylib/msgp/msgp"
+)
+
+//go:generate msgp
+
+func init() {
+	msgp.RegisterExtension(95, func() msgp.Extension { return &MetricName{} })
+	msgp.RegisterExtension(90, func() msgp.Extension { return &TagKeyValues{} })
+}
+
+type mType uint8
+
+// MType represented by a uint8
+const (
+	MTypeUndefined mType = iota
+	MTypeGauge
+	MTypeRate
+	MTypeCount
+	MTypeCounter
+	MTypeTimestamp
+)
+
+// MetricName stores the name as a []uintptr to strings interned in an object store.
+// Each word is stored as a separate interned string without the '.'. The whole name
+// can be retrieved by calling MetricName.String()
+type MetricName struct {
+	nodes []uintptr
+}
+
+// Nodes returns the []uintptr of interned string addresses
+// for the MetricName
+func (mn *MetricName) Nodes() []uintptr {
+	return mn.nodes
+}
+
+// String returns the full MetricName as a string
+// using data interned in the object store
+func (mn *MetricName) String() string {
+	if len(mn.nodes) == 0 {
+		return ""
+	}
+
+	bld := strings.Builder{}
+	return mn.string(&bld)
+}
+
+func (mn *MetricName) string(bld *strings.Builder) string {
+	// get []int of the lengths of all of the mn.Nodes
+	lns, ok := IdxIntern.Len(mn.nodes)
+	if !ok {
+		internError.Inc()
+		log.Error("idx: Failed to retrieve length of strings from interning library for MetricName")
+		return ""
+	}
+
+	// should be faster than calling IdxIntern.SetString in a tight loop
+	var tmpSz string
+	szHeader := (*reflect.StringHeader)(unsafe.Pointer(&tmpSz))
+	first, _ := IdxIntern.GetStringFromPtr(mn.nodes[0])
+	bld.WriteString(first)
+	for idx, nodePtr := range mn.nodes[1:] {
+		szHeader.Data = nodePtr
+		szHeader.Len = lns[idx+1]
+		bld.WriteString(".")
+		bld.WriteString(tmpSz)
+	}
+
+	return bld.String()
+}
+
+// setMetricName interns the MetricName in an
+// object store and stores the addresses of those strings
+// in MetricName.nodes
+func (mn *MetricName) setMetricName(name string) {
+	nodes := strings.Split(name, ".")
+	mn.nodes = make([]uintptr, len(nodes))
+	for i, node := range nodes {
+		// TODO: add error checking? Fail somehow
+		nodePtr, err := IdxIntern.AddOrGet([]byte(node), false)
+		if err != nil {
+			log.Error("idx: Failed to acquire interned string for node name: ", err)
+			internError.Inc()
+		}
+		mn.nodes[i] = nodePtr
```

Comment:
nodePtr can be invalid here. this seems quite dangerous.
i would say if we can't add any of a MetricDefinitions properties into the object store, we should fail to add it into the index (e.g. bubble up the error)
even if we omit invalid pointers, seems too risky to have incorrect metrics in the index.

Reply:
Sounds good to me